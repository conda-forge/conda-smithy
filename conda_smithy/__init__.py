try:
    from conda_smithy._version import __version__
except ImportError:
    __version__ = "0.0.0"
