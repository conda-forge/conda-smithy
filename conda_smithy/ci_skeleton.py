"""Generate skeleton for using conda-forge CI configuration files outside of
a feedstock (i.e. a outside of a forge context). This enables people to run
their CI without having to worry about how to set it up or keep it up-to-date,
by reusing the same tools that conda-forge uses for its infrastructure.
Note that your CI jobs will still execute under your organization, and not be
added to conda-forge's queue.
"""
import os
import sys

from jinja2 import Environment, FileSystemLoader

from .configure_feedstock import make_jinja_env, conda_forge_content


def _render_template(template_file, env, forge_dir, config):
    """Renders the template"""
    template = env.get_template(
        os.path.basename(template_file) + ".ci-skel.tmpl"
    )
    target_fname = os.path.join(forge_dir, template_file)
    print("Generating " + target_fname, file=sys.stderr)
    new_file_contents = template.render(**config)
    os.makedirs(os.path.dirname(target_fname), exist_ok=True)
    with open(target_fname, "w") as fh:
        fh.write(new_file_contents)


def _insert_into_gitignore(
    feedstock_directory=".",
    prefix="# conda smithy ci-skeleton start\n",
    suffix="# conda smithy ci-skeleton end\n",
):
    """Places gitignore contents into gitignore."""
    # get current contents
    fname = os.path.join(feedstock_directory, ".gitignore")
    print("Updating " + fname)
    if os.path.isfile(fname):
        with open(fname, "r") as f:
            s = f.read()
        before, _, s = s.partition(prefix)
        _, _, after = s.partition(suffix)
    else:
        before = after = ""
        dname = os.path.dirname(fname)
        if dname:
            os.makedirs(dname, exist_ok=True)
    # get new values
    gi = os.path.join(conda_forge_content, "feedstock_content", ".gitignore")
    with open(gi, "r") as f:
        s = f.read()
    new = prefix + s + suffix
    # write out the file
    with open(fname, "w") as f:
        f.write(before + new + after)
    return fname


def generate(
    package_name="pkg", feedstock_directory=".", recipe_directory="recipe"
):
    """Generates the CI skeleton."""
    forge_dir = os.path.abspath(feedstock_directory)
    env = make_jinja_env(forge_dir)
    config = dict(
        package_name=package_name,
        feedstock_directory=feedstock_directory,
        recipe_directory=recipe_directory,
    )
    # render templates
    _render_template("conda-forge.yml", env, forge_dir, config)
    _render_template(
        os.path.join(recipe_directory, "meta.yaml"), env, forge_dir, config
    )
    # update files which may exist with other content
    _insert_into_gitignore(feedstock_directory=feedstock_directory)
