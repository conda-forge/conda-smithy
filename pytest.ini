[pytest]
markers =
    legacy_circle: Test designed to run as if prior to the azure migration
    legacy_travis: Test designed to run as if prior to the azure migration
    legacy_appveyor: Test designed to run as if prior to the azure migration
    cli: CLI tests outside of test/test_cli.py