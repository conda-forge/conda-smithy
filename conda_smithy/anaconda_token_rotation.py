"""This module updates/rotates anaconda/binstar tokens.

The correct way to use this module is to call its functions via the command
line utility. The relevant one is

    conda-smithy update-anaconda-token

Note that if you are using appveyor, you will need to push the changes to the
conda-forge.yml in your feedstock to GitHub.
"""

import os
import sys
from contextlib import redirect_stderr, redirect_stdout
from typing import Optional, Union

import requests
from github import Github

from conda_smithy.utils import update_conda_forge_config


def _get_anaconda_token():
    """use this helper to enable easier patching for tests"""
    try:
        from conda_smithy.ci_register import anaconda_token

        return anaconda_token
    except ImportError:
        raise RuntimeError(
            "You must have the anaconda token defined to do token rotation!"
        )


def rotate_anaconda_token(
    user: str,
    project: str,
    feedstock_config_path: Optional[str],
    drone: bool = True,
    circle: bool = True,
    travis: bool = True,
    azure: bool = True,
    appveyor: bool = True,
    github_actions: bool = True,
    token_name: str = "BINSTAR_TOKEN",
    drone_endpoints: Union[list, tuple] = (),
):
    """Rotate the anaconda (binstar) token used by the CI providers

    All exceptions are swallowed and stdout/stderr from this function is
    redirected to `/dev/null`. Sanitized error messages are
    displayed at the end.

    If you need to debug this function, define `DEBUG_ANACONDA_TOKENS` in
    your environment before calling this function.
    """
    # we are swallong all of the logs below, so we do a test import here
    # to generate the proper errors for missing tokens
    # note that these imports cover all providers
    from .ci_register import travis_endpoint  # noqa
    from .azure_ci_utils import default_config  # noqa
    from conda_smithy.github import gh_token

    anaconda_token = _get_anaconda_token()

    if github_actions:
        gh = Github(gh_token())

    # capture stdout, stderr and suppress all exceptions so we don't
    # spill tokens
    failed = False
    err_msg = None
    with open(os.devnull, "w") as fp:
        if "DEBUG_ANACONDA_TOKENS" in os.environ:
            fpo = sys.stdout
            fpe = sys.stderr
        else:
            fpo = fp
            fpe = fp

        with redirect_stdout(fpo), redirect_stderr(fpe):
            try:
                if circle:
                    try:
                        rotate_token_in_circle(
                            user, project, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on circle!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if drone:
                    for drone_endpoint in drone_endpoints:
                        try:
                            rotate_token_in_drone(
                                user,
                                project,
                                anaconda_token,
                                token_name,
                                drone_endpoint,
                            )
                        except Exception as e:
                            if "DEBUG_ANACONDA_TOKENS" in os.environ:
                                raise e
                            else:
                                err_msg = (
                                    f"Failed to rotate token for {user}/{project}"
                                    f" on drone endpoint {drone_endpoint}!"
                                )
                                failed = True
                                raise RuntimeError(err_msg)

                if travis:
                    try:
                        rotate_token_in_travis(
                            user,
                            project,
                            feedstock_config_path,
                            anaconda_token,
                            token_name,
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on travis!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if azure:
                    try:
                        rotate_token_in_azure(
                            user, project, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on azure!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if appveyor:
                    try:
                        rotate_token_in_appveyor(
                            feedstock_config_path, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on appveyor!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if github_actions:
                    try:
                        rotate_token_in_github_actions(
                            user, project, anaconda_token, token_name, gh
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on github actions!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

            except Exception as e:
                if "DEBUG_ANACONDA_TOKENS" in os.environ:
                    raise e
                failed = True

    if failed:
        if err_msg:
            raise RuntimeError(err_msg)
        else:
            raise RuntimeError(
                f"Rotating the feedstock token in providers for {user}/{project} failed!"
                " Try the command locally with DEBUG_ANACONDA_TOKENS"
                " defined in the environment to investigate!"
            )


def rotate_token_in_circle(user, project, binstar_token, token_name):
    from conda_smithy.ci_register import circle_token

    url_template = (
        "https://circleci.com/api/v1.1/project/github/{user}/{project}/envvar{extra}?"
        "circle-token={token}"
    )

    r = requests.get(
        url_template.format(
            token=circle_token, user=user, project=project, extra=""
        )
    )
    if r.status_code != 200:
        r.raise_for_status()

    have_binstar_token = False
    for evar in r.json():
        if evar["name"] == token_name:
            have_binstar_token = True

    if have_binstar_token:
        r = requests.delete(
            url_template.format(
                token=circle_token,
                user=user,
                project=project,
                extra=f"/{token_name}",
            )
        )
        if r.status_code != 200:
            r.raise_for_status()

    data = {"name": token_name, "value": binstar_token}
    response = requests.post(
        url_template.format(
            token=circle_token, user=user, project=project, extra=""
        ),
        data,
    )
    if response.status_code != 201:
        raise ValueError(response)


def rotate_token_in_drone(
    user, project, binstar_token, token_name, drone_endpoint
):
    from conda_smithy.ci_register import drone_session

    session = drone_session(drone_endpoint)

    r = session.get(f"/api/repos/{user}/{project}/secrets")
    r.raise_for_status()
    have_binstar_token = False
    for secret in r.json():
        if token_name == secret["name"]:
            have_binstar_token = True

    if have_binstar_token:
        r = session.patch(
            f"/api/repos/{user}/{project}/secrets/{token_name}",
            json={"data": binstar_token, "pull_request": False},
        )
        r.raise_for_status()
    else:
        response = session.post(
            f"/api/repos/{user}/{project}/secrets",
            json={
                "name": token_name,
                "data": binstar_token,
                "pull_request": False,
            },
        )
        if response.status_code != 200:
            response.raise_for_status()


def rotate_token_in_travis(
    user, project, feedstock_config_path, binstar_token, token_name
):
    """update the binstar token in travis."""
    from conda_smithy.ci_register import (
        travis_endpoint,
        travis_get_repo_info,
        travis_headers,
    )

    headers = travis_headers()

    repo_info = travis_get_repo_info(user, project)
    repo_id = repo_info["id"]

    r = requests.get(
        f"{travis_endpoint}/repo/{repo_id}/env_vars",
        headers=headers,
    )
    if r.status_code != 200:
        r.raise_for_status()

    have_binstar_token = False
    ev_id = None
    for ev in r.json()["env_vars"]:
        if ev["name"] == token_name:
            have_binstar_token = True
            ev_id = ev["id"]

    data = {
        "env_var.name": token_name,
        "env_var.value": binstar_token,
        "env_var.public": "false",
    }

    if have_binstar_token:
        r = requests.patch(
            f"{travis_endpoint}/repo/{repo_id}/env_var/{ev_id}",
            headers=headers,
            json=data,
        )
        r.raise_for_status()
    else:
        r = requests.post(
            f"{travis_endpoint}/repo/{repo_id}/env_vars",
            headers=headers,
            json=data,
        )
        if r.status_code != 201:
            r.raise_for_status()

    # we remove the token in the conda-forge.yml since on travis the
    # encrypted values override any value we put in the API
    with update_conda_forge_config(feedstock_config_path) as code:
        if (
            "travis" in code
            and "secure" in code["travis"]
            and token_name in code["travis"]["secure"]
        ):
            del code["travis"]["secure"][token_name]

            if len(code["travis"]["secure"]) == 0:
                del code["travis"]["secure"]

            if len(code["travis"]) == 0:
                del code["travis"]

            print(
                "An old value of the variable %s for travis was found in the "
                "conda-forge.yml. You may need to rerender this feedstock to "
                "use the new value since encrypted secrets inserted in travis.yml "
                "files override those set in the UI/API!"
            )


def rotate_token_in_azure(user, project, binstar_token, token_name):
    from vsts.build.v4_1.models import BuildDefinitionVariable

    from conda_smithy.azure_ci_utils import (
        build_client,
        get_default_build_definition,
    )
    from conda_smithy.azure_ci_utils import default_config as config

    bclient = build_client()

    existing_definitions = bclient.get_definitions(
        project=config.project_name, name=project
    )
    if existing_definitions:
        assert len(existing_definitions) == 1
        ed = existing_definitions[0]
    else:
        raise RuntimeError(
            f"Cannot add {token_name} to a repo that is not already registerd on azure CI!"
        )

    ed = bclient.get_definition(ed.id, project=config.project_name)

    if not hasattr(ed, "variables") or ed.variables is None:
        variables = {}
    else:
        variables = ed.variables

    variables[token_name] = BuildDefinitionVariable(
        allow_override=False,
        is_secret=True,
        value=binstar_token,
    )

    build_definition = get_default_build_definition(
        user,
        project,
        config=config,
        variables=variables,
        id=ed.id,
        revision=ed.revision,
    )

    bclient.update_definition(
        definition=build_definition,
        definition_id=ed.id,
        project=ed.project.name,
    )


def rotate_token_in_appveyor(feedstock_config_path, binstar_token, token_name):
    from conda_smithy.ci_register import appveyor_token

    headers = {"Authorization": f"Bearer {appveyor_token}"}
    url = "https://ci.appveyor.com/api/account/encrypt"
    response = requests.post(
        url, headers=headers, data={"plainValue": binstar_token}
    )
    if response.status_code != 200:
        raise ValueError(response)

    with update_conda_forge_config(feedstock_config_path) as code:
        code.setdefault("appveyor", {}).setdefault("secure", {})[
            token_name
        ] = response.content.decode("utf-8")


def rotate_token_in_github_actions(
    user, project, binstar_token, token_name, gh
):
    repo = gh.get_repo(f"{user}/{project}")
    assert repo.create_secret(token_name, binstar_token)
