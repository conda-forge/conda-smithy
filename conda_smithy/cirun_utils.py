"""
See http://py.cirun.io/api.html for cirun client docs
"""

from functools import lru_cache
from typing import Any, Dict, List, Optional

from cirun import Cirun

from conda_smithy.github import Github, gh_token


@lru_cache
def get_cirun_installation_id(owner: str) -> int:
    # This ID needs a token with admin: org privileges.
    # Hard-code instead for easier use.
    if owner == "conda-forge":
        return 18453316
    else:
        gh = Github(gh_token())
        user = gh.get_user()
        user_or_org: Any
        if user.login == owner:
            user_or_org = user
        else:
            user_or_org = gh.get_organization(owner)
        for inst in user_or_org.get_installations():
            if inst.raw_data["app_slug"] == "cirun-application":
                return inst.app_id
        raise ValueError(f"cirun not found for owner {owner}")


def enable_cirun_for_project(owner: str, repo: str) -> Dict[str, Any]:
    """Enable the cirun.io Github Application for a particular repository."""
    print(f"Enabling cirun for {owner}/{repo} ...")
    cirun = _get_cirun_client()
    return cirun.set_repo(
        f"{owner}/{repo}", installation_id=get_cirun_installation_id(owner)
    )


def add_repo_to_cirun_resource(
    owner: str,
    repo: str,
    resources: List[str],
    teams: Optional[List],
    roles: Optional[List],
    users_from_json: Optional[str] = None,
    cirun_policy_args: Optional[List[str]] = None,
) -> Dict[str, Any]:
    """Grant access to a cirun resource to a particular repository, with a particular policy."""
    cirun = _get_cirun_client()
    policy_args: Optional[Dict[str, Any]] = None
    if cirun_policy_args and "pull_request" in cirun_policy_args:
        policy_args = {"pull_request": True}
    print(
        f"Adding repo {owner}/{repo} to resources {resources} with policy_args: {policy_args}"
    )
    gh = Github(gh_token())
    gh_owner = gh.get_user(owner)
    gh_repo = gh_owner.get_repo(repo)

    # Need to send None instead of an empty list
    teams = teams or None
    roles = roles or None
    if not (teams or roles or users_from_json):
        teams = [team.name for team in gh_repo.get_teams()]

    response = cirun.add_repo_to_resources(
        owner,
        repo,
        resources=resources,
        policy_args=policy_args,
        teams=teams,
        roles=roles,
        users_from_json=users_from_json,
    )
    print(f"response: {response} | {response.json().keys()}")
    return response


def remove_repo_from_cirun_resource(owner: str, repo: str, resource: str):
    """Revoke access to a cirun resource to a particular repository, with a particular policy."""
    cirun = _get_cirun_client()
    print(f"Removing repo {owner}/{repo} from resource {resource}.")
    response = cirun.remove_repo_from_resources(owner, repo, [resource])
    print(f"response: {response} | {response.json().keys()}")
    return response


@lru_cache
def _get_cirun_client() -> Cirun:
    try:
        return Cirun()
    except KeyError:
        raise RuntimeError(
            "You must have CIRUN_API_KEY defined to do Cirun CI registration. "
            "This requirement can be overriden by specifying `--without-cirun`"
        )
