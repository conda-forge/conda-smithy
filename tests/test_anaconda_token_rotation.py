from unittest import mock

import pytest

from conda_smithy.anaconda_token_rotation import rotate_anaconda_token
from conda_smithy.ci_register import drone_default_endpoint


@pytest.mark.parametrize("appveyor", [True, False])
@pytest.mark.parametrize("drone", [True, False])
@pytest.mark.parametrize("circle", [True, False])
@pytest.mark.parametrize("azure", [True, False])
@pytest.mark.parametrize("travis", [True, False])
@pytest.mark.parametrize("github_actions", [True, False])
@mock.patch("conda_smithy.github.gh_token")
@mock.patch("conda_smithy.anaconda_token_rotation._get_anaconda_token")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_appveyor")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_drone")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_circle")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_travis")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_azure")
@mock.patch(
    "conda_smithy.anaconda_token_rotation.rotate_token_in_github_actions"
)
def test_rotate_anaconda_token(
    github_actions_mock,
    azure_mock,
    travis_mock,
    circle_mock,
    drone_mock,
    appveyor_mock,
    get_ac_token,
    get_gh_token,
    appveyor,
    drone,
    circle,
    azure,
    travis,
    github_actions,
):
    user = "foo"
    project = "bar"

    anaconda_token = "abc123"
    get_ac_token.return_value = anaconda_token
    get_gh_token.return_value = None

    feedstock_config_path = "abc/conda-forge.yml"

    rotate_anaconda_token(
        user,
        project,
        feedstock_config_path,
        drone=drone,
        circle=circle,
        travis=travis,
        azure=azure,
        appveyor=appveyor,
        github_actions=github_actions,
        token_name="MY_FANCY_TOKEN",
        drone_endpoints=[drone_default_endpoint],
    )

    if drone:
        drone_mock.assert_called_once_with(
            user,
            project,
            anaconda_token,
            "MY_FANCY_TOKEN",
            drone_default_endpoint,
        )
    else:
        drone_mock.assert_not_called()

    if circle:
        circle_mock.assert_called_once_with(
            user, project, anaconda_token, "MY_FANCY_TOKEN"
        )
    else:
        circle_mock.assert_not_called()

    if travis:
        travis_mock.assert_called_once_with(
            user,
            project,
            feedstock_config_path,
            anaconda_token,
            "MY_FANCY_TOKEN",
        )
    else:
        travis_mock.assert_not_called()

    if azure:
        azure_mock.assert_called_once_with(
            user, project, anaconda_token, "MY_FANCY_TOKEN"
        )
    else:
        azure_mock.assert_not_called()

    if appveyor:
        appveyor_mock.assert_called_once_with(
            feedstock_config_path, anaconda_token, "MY_FANCY_TOKEN"
        )
    else:
        appveyor_mock.assert_not_called()

    if github_actions:
        github_actions_mock.assert_called_once_with(
            user,
            project,
            anaconda_token,
            "MY_FANCY_TOKEN",
            mock.ANY,
        )
    else:
        github_actions_mock.assert_not_called()


@pytest.mark.parametrize("appveyor", [True, False])
@pytest.mark.parametrize("drone", [True, False])
@pytest.mark.parametrize("circle", [True, False])
@pytest.mark.parametrize("azure", [True, False])
@pytest.mark.parametrize("travis", [True, False])
@pytest.mark.parametrize("github_actions", [True, False])
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_appveyor")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_drone")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_circle")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_travis")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_azure")
@mock.patch(
    "conda_smithy.anaconda_token_rotation.rotate_token_in_github_actions"
)
def test_rotate_anaconda_token_notoken(
    github_actions_mock,
    azure_mock,
    travis_mock,
    circle_mock,
    drone_mock,
    appveyor_mock,
    appveyor,
    drone,
    circle,
    azure,
    travis,
    github_actions,
    monkeypatch,
):
    user = "foo"
    project = "bar"

    with pytest.raises(RuntimeError) as e:
        rotate_anaconda_token(
            user,
            project,
            None,
            drone=drone,
            circle=circle,
            travis=travis,
            azure=azure,
            appveyor=appveyor,
            github_actions=github_actions,
            drone_endpoints=[drone_default_endpoint],
        )

    assert "anaconda token" in str(e.value)

    drone_mock.assert_not_called()
    circle_mock.assert_not_called()
    travis_mock.assert_not_called()
    azure_mock.assert_not_called()
    appveyor_mock.assert_not_called()
    github_actions_mock.assert_not_called()


@pytest.mark.parametrize(
    "provider",
    ["drone", "circle", "travis", "azure", "appveyor", "github_actions"],
)
@mock.patch("conda_smithy.github.gh_token")
@mock.patch("conda_smithy.anaconda_token_rotation._get_anaconda_token")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_appveyor")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_drone")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_circle")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_travis")
@mock.patch("conda_smithy.anaconda_token_rotation.rotate_token_in_azure")
@mock.patch(
    "conda_smithy.anaconda_token_rotation.rotate_token_in_github_actions"
)
def test_rotate_anaconda_token_provider_error(
    github_actions_mock,
    azure_mock,
    travis_mock,
    circle_mock,
    drone_mock,
    appveyor_mock,
    get_ac_token,
    get_gh_token,
    provider,
):
    user = "foo"
    project = "bar"

    anaconda_token = "abc123"
    get_ac_token.return_value = anaconda_token
    get_gh_token.return_value = None

    user = "foo"
    project = "bar-feedstock"

    if provider == "drone":
        drone_mock.side_effect = ValueError("blah")
    if provider == "circle":
        circle_mock.side_effect = ValueError("blah")
    if provider == "travis":
        travis_mock.side_effect = ValueError("blah")
    if provider == "azure":
        azure_mock.side_effect = ValueError("blah")
    if provider == "appveyor":
        appveyor_mock.side_effect = ValueError("blah")
    if provider == "github_actions":
        github_actions_mock.side_effect = ValueError("blah")

    with pytest.raises(RuntimeError) as e:
        rotate_anaconda_token(
            user, project, None, drone_endpoints=[drone_default_endpoint]
        )

    assert "on {}".format(provider.replace("_", " ")) in str(e.value)
