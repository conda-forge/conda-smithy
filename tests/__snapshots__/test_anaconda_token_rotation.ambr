# serializer version: 1
# name: test_rotate_anaconda_token_notoken[False-False-False-False-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-False-False-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-False-False-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-False-False-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-False-True-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-False-True-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-False-True-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-False-True-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-True-False-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-True-False-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-True-False-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-True-False-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-True-True-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-True-True-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-True-True-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-False-True-True-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-False-False-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-False-False-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-False-False-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-False-False-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-False-True-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-False-True-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-False-True-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-False-True-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-True-False-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-True-False-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-True-False-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-True-False-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-True-True-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-True-True-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-True-True-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[False-True-True-True-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-False-False-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-False-False-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-False-False-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-False-False-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-False-True-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-False-True-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-False-True-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-False-True-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-True-False-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-True-False-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-True-False-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-True-False-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-True-True-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-True-True-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-True-True-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-False-True-True-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-False-False-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-False-False-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-False-False-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-False-False-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-False-True-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-False-True-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-False-True-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-False-True-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-True-False-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-True-False-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-True-False-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-True-False-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-True-True-False-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-True-True-False-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-True-True-True-False]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_notoken[True-True-True-True-True-True]
  'You must have the anaconda token defined to do token rotation!'
# ---
# name: test_rotate_anaconda_token_provider_error[appveyor]
  'Failed to rotate token for foo/bar-feedstock on appveyor!'
# ---
# name: test_rotate_anaconda_token_provider_error[azure]
  'Failed to rotate token for foo/bar-feedstock on azure!'
# ---
# name: test_rotate_anaconda_token_provider_error[circle]
  'Failed to rotate token for foo/bar-feedstock on circle!'
# ---
# name: test_rotate_anaconda_token_provider_error[drone]
  'Failed to rotate token for foo/bar-feedstock on drone endpoint https://cloud.drone.io!'
# ---
# name: test_rotate_anaconda_token_provider_error[github_actions]
  'Failed to rotate token for foo/bar-feedstock on github actions!'
# ---
# name: test_rotate_anaconda_token_provider_error[travis]
  'Failed to rotate token for foo/bar-feedstock on travis!'
# ---
